ocaml/xapi/xapi_vm_migrate.ml

```diff
@@ -904,14 +913,22 @@ let migrate_send'  ~__context ~vm ~dest ~live ~vdi_map ~vif_map ~options =
         | Xenops_interface.Does_not_exist ("extra",_) ->
           ()
       end;
-      detach_local_network_for_vm ~__context ~vm ~destination:remote.dest_host;
+
       debug "Migration complete";
       SMPERF.debug "vm.migrate_send: migration complete vm:%s" vm_uuid;
 
-      Xapi_xenops.refresh_vm ~__context ~self:vm;
-
+      (* So far the main body of migration is completed, and the rests are
+         updates, config or cleanup on the source and destination. There will
+         be no (clean) way back from this point, due to these destructive
+         changes, so we don't want user intervention e.g. task cancellation.
+       *)
+      TaskHelper.exn_if_cancelling ~__context;
+      TaskHelper.set_not_cancellable ~__context;
       XenAPI.VM.pool_migrate_complete remote.rpc remote.session new_vm remote.dest_host;
 
+      detach_local_network_for_vm ~__context ~vm ~destination:remote.dest_host;
+      Xapi_xenops.refresh_vm ~__context ~self:vm;
```

Comment:
The `detach_local_network_vm` are destructive post-migration change, so better leave together to the after completion section (where task cancellation is now disabled for). 

Bringing the `pool_migration_complete` a little bit forward (i.e. immediately after the domain migration) seems a little bit safer to me, hence the change. Nonetheless, there shouldn't be much interference between this (remote) op and the two (local) ops  with a slightly different orders.